File: examples/footguns/04_task_leak/main.c
/*
 * Footgun: Task Leak — C
 *
 * A detached pthread blocks indefinitely on a mutex it will never
 * acquire. The main thread exits, but the leaked thread and its
 * resources (stack, TLS, mutex state) persist until the OS cleans up
 * the process. In a long-running server that spawns threads per-request,
 * this exhausts the thread limit and eventually deadlocks the server.
 *
 * C has no structured concurrency. pthread_detach() is "fire and forget"
 * by design. Nothing warns you that the thread will block forever.
 */

#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

static pthread_mutex_t held = PTHREAD_MUTEX_INITIALIZER;

static void *blocked_worker(void *arg) {
    (void)arg;
    /* This lock is held by main and never released.
     * This thread blocks here until the process exits. */
    pthread_mutex_lock(&held);
    printf("worker: got lock (never reached)\n");
    pthread_mutex_unlock(&held);
    return NULL;
}

int main(void) {
    /* Hold the mutex so the worker blocks forever. */
    pthread_mutex_lock(&held);

    for (int i = 0; i < 5; i++) {
        pthread_t t;
        pthread_create(&t, NULL, blocked_worker, NULL);
        pthread_detach(t); /* detached: no join, no cleanup signal */
        printf("spawned detached thread %d\n", i + 1);
    }

    printf("main: exiting with 5 leaked threads still blocked\n");
    /* pthread_mutex_unlock(&held); -- never called */
    return 0;
}

/*
 * Compile and run:
 *   gcc -pthread -o task_leak main.c && ./task_leak
 *
 * Result: main exits, OS reclaims everything — but in a daemon process
 * this accumulates until ulimit -u is hit. No runtime warning.
 *
 * Fix: use pthread_cancel() + pthread_join(), or pass a shutdown flag
 * and a condition variable so workers can exit cleanly.
 */

File: examples/footguns/04_task_leak/main.rs
// Footgun: Task Leak — Rust
//
// Rust's thread::spawn returns a JoinHandle. Dropping the handle
// detaches the thread — it runs to completion but is no longer
// reachable for joining. If the thread blocks indefinitely, it leaks.
// This is safe in Rust's memory sense (no UB), but it is a resource
// leak with the same operational consequences as in C or Go.
//
// Rust does not prevent task leaks. The type system only ensures
// memory safety, not liveness. thread::spawn is "fire and forget"
// if you drop the JoinHandle.

use std::sync::{Arc, Condvar, Mutex};
use std::thread;

fn spawn_blocked() -> thread::JoinHandle<()> {
    // This condvar is never notified; the thread blocks forever.
    let pair = Arc::new((Mutex::new(false), Condvar::new()));
    let pair2 = Arc::clone(&pair);

    thread::spawn(move || {
        let (lock, cvar) = &*pair2;
        let mut ready = lock.lock().unwrap();
        while !*ready {
            ready = cvar.wait(ready).unwrap(); // blocks forever
        }
        println!("notified (never reached)");
    })
}

fn handle_request(id: u32) {
    let handle = spawn_blocked();
    drop(handle); // detaches the thread — it leaks
    println!("request {} handled (thread leaked)", id);
}

fn main() {
    for i in 0..5 {
        handle_request(i);
    }
    println!("main exiting with 5 leaked threads");
    // OS reclaims on process exit, but in a daemon this accumulates.
}

// Compile and run:  rustc main.rs -o task_leak && ./task_leak
//
// Rust's async ecosystem has the same problem:
//   tokio::spawn(async { /* blocks forever */ });
//   // JoinHandle dropped → task is detached → leaks
//
// Fix (threads): retain and join the JoinHandle, or use a thread pool
// with a shutdown mechanism (e.g. crossbeam::thread::scope).
//
// Fix (async): use structured concurrency via tokio::task::JoinSet
// or async-std's task groups, which cancel children on drop.
